File: m3meka/python/m3/joint_array.py
#M3 -- Meka Robotics Robot Components
#Copyright (c) 2010 Meka Robotics
#Author: edsinger@mekabot.com (Aaron Edsinger)

#M3 is free software: you can redistribute it and/or modify
#it under the terms of the GNU Lesser General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#M3 is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Lesser General Public License for more details.

#You should have received a copy of the GNU Lesser General Public License
#along with M3.  If not, see <http://www.gnu.org/licenses/>.

import yaml
import os 
import m3.toolbox as m3t
import m3.toolbox_ctrl as m3tc
import m3.joint_array_pb2 as mab
import m3.joint_array_mode_pb2 as mam
import m3.smoothing_mode_pb2 as msm
import m3.joint_pb2 as mrt
from m3.component import M3Component
from m3.unit_conversion import *

import scipy.linalg
import numpy as nu


class M3JointArray(M3Component):
    """Interface for a joint array.
    The kinematic data and sensor data from all DOF is loaded into Numeric arrays.
    on each update from the server. 
    If not specified, all units are meters/mNm/radians/seconds.

    The set_xxx commands support a number of different forms:
    For example:
    -- set_pwm(a) : sets all joints to a
    -- set_pwm(a,[1,2,3]), set joints 1,2,3 to a
    -- set_pwm(a,1), set joint 1 to a
    -- set_pwm([a b c],[1,2,3]) set joints 1,2,3 to a,b,c
    -- set_pwm([a,b,...]): sets all joints to a,b,...
    """

    def __init__(self,name,ndof,ctype):
        M3Component.__init__(self,name,ctype)
	self.read_config()
        self.status=mab.M3JointArrayStatus()
        self.command=mab.M3JointArrayCommand()
        self.param=mab.M3JointArrayParam()
	if ndof is None:
	    ndof=self.config['ndof']
        self.ndof=ndof
        self.vias=[]
        self.via_idx=0
        # Grow the protocol buffer messages to correct size
        # NOTE: should really support mixed sizes but...
        for i in range(self.ndof):
            self.command.pwm_desired.append(0)
            self.command.tq_desired.append(0)
            self.command.q_desired.append(0)
            self.command.qdot_desired.append(0)
            self.command.q_stiffness.append(0)
	    self.command.q_slew_rate.append(0)
	    self.command.smoothing_mode.append(msm.SMOOTHING_MODE_OFF)
            self.command.ctrl_mode.append(mam.JOINT_ARRAY_MODE_OFF)
        
        self.motor_temp=nu.zeros(self.ndof,float)
        self.amp_temp=nu.zeros(self.ndof,float)
        self.torque=nu.zeros(self.ndof,float)
        self.torquedot=nu.zeros(self.ndof,float)	
        self.current=nu.zeros(self.ndof,float)
        self.theta=nu.zeros(self.ndof,float)
        self.thetadot=nu.zeros(self.ndof,float)
        self.thetadotdot=nu.zeros(self.ndof,float)
	
	self.max_slew_rates=self.__get_max_slew_rates_from_config()

    
    def get_limb_name(self): 
        """Unique limb name: 'left_arm','right_arm',etc"""
        return self.config['limb_name'] 

    def get_num_dof(self):
        return self.ndof    

    def get_motor_temp_C(self): 
        return self.motor_temp
    def get_motor_temp_F(self): 
        return C2F(self.motor_temp)
    def get_amp_temp_C(self): 
        return self.amp_temp
    def get_amp_temp_F(self): 
        return C2F(self.amp_temp)
    def get_current_mA(self): 
        return self.current
    def get_total_current_mA(self):
        return nu.sum(self.current)

    def get_torque_mNm(self): 
        return self.torque
    def get_torque_inLb(self): 
        return mNm2inLb(self.torque)
    def get_torquedot_mNm(self): 
        return self.torquedot
    def get_torquedot_inLb(self): 
        return mNm2inLb(self.torquedot)       
    def get_theta_deg(self): 
        return self.theta
    def get_theta_rad(self): 
        return deg2rad(self.theta)
    def get_thetadot_rad(self): 
        return deg2rad(self.thetadot)
    def get_thetadot_deg(self): 
        return self.thetadot
    def get_thetadotdot_rad(self): 
        return deg2rad(self.thetadotdot)
    def get_thetadotdot_deg(self): 
        return self.thetadotdot
    def get_timestamp_uS(self):
        return self.status.base.timestamp
    
    def get_pwm_cmd(self):
        return self.status.pwm_cmd
    def get_flags(self):
        return self.status.flags
    def get_limitswitch_pos(self):
	return [ int((x & aepb.ACTUATOR_EC_FLAG_POS_LIMITSWITCH)!=0) for x in self.status.flags]
    def get_limitswitch_neg(self):
	return [int((x & aepb.ACTUATOR_EC_FLAG_NEG_LIMITSWITCH)!=0) for x in self.status.flags]
    def get_encoder_calibrated(self):
	return [int((x & aepb.ACTUATOR_EC_FLAG_QEI_CALIBRATED)!=0) for x in self.status.flags]
    
    def set_pwm(self,v,ind=None):
        M3Component.set_int_array(self,self.command.pwm_desired,v,ind)
    def set_torque_mNm(self,v,ind=None):
        M3Component.set_float_array(self,self.command.tq_desired,v,ind)
    def set_torque_inLb(self,v,ind=None):
        M3Component.set_float_array(self,self.command.tq_desired,inLb2mNm(nu.array(v)),ind)
    def set_theta_rad(self,v,ind=None):
        M3Component.set_float_array(self,self.command.q_desired,rad2deg(nu.array(v)),ind)
    def set_theta_deg(self,v,ind=None):
        M3Component.set_float_array(self,self.command.q_desired,v,ind)
    def set_stiffness(self,v,ind=None):
        M3Component.set_float_array(self,self.command.q_stiffness,v,ind)
    def set_slew_rate(self,v):
        M3Component.set_float_array(self,self.command.q_slew_rate,v,ind)
    def set_slew_rate_proportion(self,v,ind=None):
	slew_rates = []
	if ind is not None:
	    for i in range(len(ind)):
	    	slew_rates.append(self.max_slew_rates[ind[i]]*max(0.0,min(1.0,v[i])))
	else:
	    for i in range(len(v)):
		slew_rates.append(self.max_slew_rates[i]*max(0.0,min(1.0,v[i])))	
	M3Component.set_float_array(self,self.command.q_slew_rate,slew_rates,ind)
    def set_mode(self,v,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,v,ind)
    def set_mode_off(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_OFF,ind)
    def set_mode_pwm(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_PWM,ind)
    def set_mode_torque(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_TORQUE,ind)
    def set_mode_torque_gc(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_TORQUE_GC,ind)
    def set_mode_theta(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_THETA,ind)
    def set_mode_theta_gc(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_THETA_GC,ind)
    def set_mode_theta_mj(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_THETA_MJ,ind)
    def set_mode_theta_gc_mj(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_THETA_GC_MJ,ind)
    def set_mode_splined_traj(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_SPLINED_TRAJ,ind)
    def set_mode_splined_traj_gc(self,ind=None):
        M3Component.set_int_array(self,self.command.ctrl_mode,mam.JOINT_ARRAY_MODE_SPLINED_TRAJ_GC,ind)

    def add_splined_traj_via_rad(self,theta_des,thetadot_avg):
        """Add a desired via point to the queue. 
	-- thetadot_avg is the desired average velocity of the joint with the farthest to travel
	Requires that a NDOF are specified even if not all are in via control mode"""
        self.vias.append([rad2deg(nu.array(theta_des,float)),rad2deg(nu.array(thetadot_avg))])

    def add_splined_traj_via_deg(self,theta_des,thetadot_avg):
        self.vias.append([theta_des,thetadot_avg])

    def is_splined_traj_complete(self):
        """Have all the joint vias finished executing on the server"""
        return len(self.vias)==0 and self.via_idx==self.status.completed_spline_idx

    def load_command(self):
        self.command.ClearField('vias')
        nadd=min(20,len(self.vias)) #only add 20 per cycle to keep packet size down
        for n in range(nadd):
            self.via_idx=self.via_idx+1
            theta_des=self.vias[n][0]
            thetadot_avg=self.vias[n][1]
            self.command.vias.add()
            for i in range(self.ndof):
                self.command.vias[-1].q_desired.append(float(theta_des[i]))
                self.command.vias[-1].qdot_avg.append(float(thetadot_avg[i]))
            self.command.vias[-1].idx=self.via_idx
        self.vias=self.vias[nadd:]

    def update_status(self):
        self.motor_temp=nu.array(self.status.motor_temp,float)
        self.amp_temp=  nu.array(self.status.amp_temp,float)
        self.current=  nu.array(self.status.current,float)
        self.torque=  nu.array(self.status.torque,float)
        self.torquedot=  nu.array(self.status.torquedot,float)	
        self.theta=  nu.array(self.status.theta,float)
        self.thetadot=  nu.array(self.status.thetadot,float)
        self.thetadotdot=  nu.array(self.status.thetadotdot,float)

    def __get_max_slew_rates_from_config(self):
	file_name = m3t.get_component_config_filename(self.name)
	try:
            f=file(file_name,'r')
            config= yaml.safe_load(f.read())
        except (IOError, EOFError):
            print 'Config file not present:',file_name
            return
	max_slew_rates = [0.0]*self.get_num_dof()
	for k in config['joint_components']:
	    joint_name = config['joint_components'][k]
	    joint_file_name = m3t.get_component_config_filename(joint_name)
	    try:
		jf=file(joint_file_name,'r')
		joint_config= yaml.safe_load(jf.read())
	    except (IOError, EOFError):
		print 'Config file not present:', joint_file_name
		return	    
	    max_slew_rates[int(k[1:])] = joint_config['param']['max_q_slew_rate']
	return max_slew_rates

File: m3meka/src/m3/chains/joint_array.cpp
/* 
M3 -- Meka Robotics Robot Components
Copyright (c) 2010 Meka Robotics
Author: edsinger@mekabot.com (Aaron Edsinger)

M3 is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

M3 is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with M3.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "m3/chains/joint_array.h"
#include "m3rt/base/component_factory.h"
#include "m3rt/base/toolbox.h"
#include "m3rt/base/m3rt_def.h"

namespace m3
{
	
using namespace m3rt;
using namespace std;
using namespace KDL;


///////////////////////////////////////////////////////
//Success so long as least one dependent comp. available
bool M3JointArray::LinkDependentComponents()
{	
	int nl=0;
	if (joint_names.size()>0)
		joints.assign(joint_names.size(),(M3Joint*)NULL);
	for (int i=0;i<joint_names.size();i++)
	{
		if(joint_names[i].size())
		{
			joints[i]=(M3Joint*)factory->GetComponent(joint_names[i]);			
			if (joints[i]!=NULL)
				nl++;
		}
		else
			joints[i]=NULL;
	}	
		
	if (nl<=0)
	{
		M3_ERR("M3JointArray %s found %d joints \n",GetName().c_str(),nl);	
		return false;
	}
	if (nl!=ndof)
	{
		M3_ERR("M3JointArray %s found %d joints. Expected %d. Continuing with missing joints. \n",
		       GetName().c_str(),nl,ndof);	
	}
	return true;
}

void M3JointArray::Startup()
{		
	joints.resize(ndof);
	traj_active.assign(ndof,false);
	joint_names.resize(ndof);
	traj_des.resize(ndof);
	theta.resize(ndof);
	thetadot.resize(ndof);
	thetadotdot.resize(ndof);
	torque.resize(ndof);
	tw_torque.resize(ndof);

	for (int i=0;i<ndof;i++)
	{
		if (joint_names[i].size()==0)
		{
			M3_WARN("M3JointArray joint %d not configured properly\n",i);
			SetStateError();
			return;
		}
		command.add_q_desired(0);
		command.add_qdot_desired(0);
		command.add_pwm_desired(0);
		command.add_tq_desired(0);
		command.add_q_stiffness(0);
		command.add_q_slew_rate(0);
		command.add_ctrl_mode(JOINT_ARRAY_MODE_OFF);
		command.add_smoothing_mode(SMOOTHING_MODE_OFF);
		
		status.add_motor_temp(0);
		status.add_amp_temp(0);
		status.add_current(0);
		status.add_torque(0);
		status.add_torquedot(0);		
		status.add_theta(0);
		status.add_thetadot(0);
		status.add_thetadotdot(0);
		status.add_pwm_cmd(0);
		status.add_flags(0);
	}
	status.set_completed_spline_idx(-1);
	
	SetStateSafeOp();
	tmp_cnt = 0;
}


void M3JointArray::Shutdown()
{

}


void M3JointArray::StepJointTrajectory()
{
	bool reset=false; //Check for change in active DOF
	//See if active joints has changed
	for (int i=0;i<ndof;i++)
	{
		bool st = command.ctrl_mode(i)==JOINT_ARRAY_MODE_SPLINED_TRAJ;
		bool stg= command.ctrl_mode(i)==JOINT_ARRAY_MODE_SPLINED_TRAJ_GC;
		if ( ( (st||stg) && !traj_active[i]) || ((!st&&!stg) && traj_active[i]))
			reset=true;
		traj_active[i]=(st||stg);
	}
	if (reset)
	{
		traj.Reset(traj_active,ndof);
	}
	//Add vias
	for (int i=0;i<command.vias_size();i++)
		traj.AddVia(command.vias(i));
	command.clear_vias();
	
	for (int i=0;i<ndof;i++)
	{
		theta(i)=status.theta(i);
		thetadot(i)=status.thetadot(i);
	}
	int idx=traj.Step(status.base().timestamp(),theta,thetadot,traj_des);
	status.set_completed_spline_idx(idx);
}


void M3JointArray::StepCommand()
{	
	if (IsStateSafeOp() || IsStateError())
		return;
	
	StepJointTrajectory();

	//Individual controllers
	for (int i=0;i<ndof;i++)
	{

		if (joints[i]!=NULL)
		{		
			joints[i]->SetDesiredThetaDeg(command.q_desired(i));
			joints[i]->SetDesiredThetaDotDeg(command.qdot_desired(i));
			joints[i]->SetDesiredPwm(command.pwm_desired(i));
			joints[i]->SetDesiredTorque(command.tq_desired(i));
			joints[i]->SetDesiredStiffness(command.q_stiffness(i));
			joints[i]->SetSlewRate(command.q_slew_rate(i));
			joints[i]->SetDesiredSmoothingMode(command.smoothing_mode(i));
			
			/*if (i == 2 && tmp_cnt++ == 200)
			{
				M3_DEBUG("mo: %d\n", (int)command.ctrl_mode(i));
				M3_DEBUG("th: %f\n", command.q_desired(i));
				M3_DEBUG("st: %f\n", command.q_stiffness(i));
				M3_DEBUG("sm: %d\n", (int)command.smoothing_mode(i));
				M3_DEBUG("sl: %f\n", command.q_slew_rate(i));
				M3_DEBUG("qd: %f\n", command.qdot_desired(i));
				M3_DEBUG("tq: %f\n", command.tq_desired(i));
				tmp_cnt = 0;
			}*/
			switch(command.ctrl_mode(i))
			{										
				case JOINT_ARRAY_MODE_OFF:
					joints[i]->SetDesiredControlMode(JOINT_MODE_OFF);
					break;
				case JOINT_ARRAY_MODE_PWM:
					joints[i]->SetDesiredControlMode(JOINT_MODE_PWM);
					break;
				case JOINT_ARRAY_MODE_TORQUE:
					joints[i]->SetDesiredControlMode(JOINT_MODE_TORQUE);
					break;
				case JOINT_ARRAY_MODE_THETA:

					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA);
					break;
				case JOINT_ARRAY_MODE_TORQUE_GC:
					joints[i]->SetDesiredControlMode(JOINT_MODE_TORQUE_GC);
					break;
				case JOINT_ARRAY_MODE_THETA_GC:					
					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA_GC);
					break;
				case JOINT_ARRAY_MODE_THETA_MJ:
					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA_MJ);
					break;
				case JOINT_ARRAY_MODE_THETA_GC_MJ:
					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA_GC_MJ);
					break;
				case JOINT_ARRAY_MODE_SPLINED_TRAJ_GC:
				{
					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA_GC);	
					joints[i]->SetDesiredThetaDeg(traj_des(i));
					break;
				}
				case JOINT_ARRAY_MODE_SPLINED_TRAJ:
				{
					joints[i]->SetDesiredControlMode(JOINT_MODE_THETA);	
					joints[i]->SetDesiredThetaDeg(traj_des(i));
					break;
				}
				default:
					joints[i]->SetDesiredControlMode(JOINT_MODE_OFF);
			};
		}
	}
}

void M3JointArray::StepStatus()
{	
	if (IsStateError())
		return;
	for (int i=0;i<ndof;i++)
	{
		if (joints[i]!=NULL)
		{
			status.set_motor_temp(i,joints[i]->GetMotorTemp());
			status.set_amp_temp(i,joints[i]->GetAmpTemp());
			status.set_current(i,joints[i]->GetCurrent());
			status.set_torque(i,joints[i]->GetTorque());
			status.set_torquedot(i,joints[i]->GetTorqueDot());
			status.set_torquedot(i,joints[i]->GetTorqueDot());
			status.set_theta(i,joints[i]->GetThetaDeg());
			status.set_thetadot(i,joints[i]->GetThetaDotDeg());
			status.set_thetadotdot(i,joints[i]->GetThetaDotDotDeg());
			status.set_pwm_cmd(i,joints[i]->GetPwmCmd());
			status.set_flags(i,joints[i]->GetFlags());
			torque(i)=status.torque(i);
			theta(i)=status.theta(i);
			thetadot(i)=status.thetadot(i);
			thetadotdot(i)=status.thetadotdot(i);
		}
	}
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////

bool M3JointArray::ReadConfig(const char * filename)
{
	if (!M3Component::ReadConfig(filename))
		return false;
	YAML::Node doc;
	GetYamlDoc(filename, doc);
	doc["ndof"] >> ndof;
	for(YAML::Iterator it=doc["joint_components"].begin();it!=doc["joint_components"].end();++it) 
	{
   		string key, value;
    		it.first() >> key;
    		it.second() >> value;
		int id=atoi(key.substr(1).c_str()); //"J0" gives 0, etc
		if (joint_names.size() <= (id+1))
		{
			joint_names.resize(id+1);
		}
		joint_names[id]=value;
	}
	
	return true;
}




}
